core/domain/stats_jobs_one_off_test.py

```diff
@@ -0,0 +1,3630 @@
+    """Tests for the answer migration job."""
+
+    DEMO_EXP_ID = '16'
+    DEFAULT_RULESPEC_STR = 'Default'
+
+    # This is based on the old stats_models.process_submitted_answer().
+    def _record_old_answer(
+            self, state_name, rule_spec_str, answer_html_str,
+            exploration_id=DEMO_EXP_ID, submitted_answer_count=1):
+        answer_log = stats_models.StateRuleAnswerLogModel.get_or_create(
+            exploration_id, state_name, rule_spec_str)
+        if answer_html_str in answer_log.answers:
+            answer_log.answers[answer_html_str] += submitted_answer_count
+        else:
+            answer_log.answers[answer_html_str] = submitted_answer_count
+        try:
+            answer_log.put()
+        except Exception as e:
+            logging.error(e)
+
+    def _get_split_entity_count_for_answer_log_entity(
+            self, state_name, rule_spec_str, exploration_id=DEMO_EXP_ID):
+        item_id = '.'.join(
+            [exploration_id, state_name, 'submit', rule_spec_str])[:490]
+        return (
+            stats_models.LargeAnswerBucketModel.get_split_entity_count_for_answer_log_entity(item_id)) # pylint: disable=line-too-long
+
+    def _get_submitted_answer_dict_list(self, state_answers):
+        submitted_answer_dict_list = (
+            state_answers.get_submitted_answer_dict_list())
+        for submitted_answer_dict in submitted_answer_dict_list:
+            del submitted_answer_dict['json_size']
+        return submitted_answer_dict_list
+
+    def _run_split_large_answer_buckets_job(self):
+        """Start the SplitLargeAnswerBucketsJob to possibly split up answers in
+        StateRuleAnswerLogModel.
+        """
+        job_id = stats_jobs_one_off.SplitLargeAnswerBucketsJob.create_new()
+        stats_jobs_one_off.SplitLargeAnswerBucketsJob.enqueue(job_id)
+        self.process_and_flush_pending_tasks()
+        return jobs.get_job_output(job_id)
+
+    def _run_cleanup_large_bucket_labels_from_new_answers_job(self):
+        """Start the CleanupLargeBucketLabelsFromNewAnswersJob to cleanup large
+        bucket IDs in newly stored answers.
+        """
+        job_id = stats_jobs_one_off.CleanupLargeBucketLabelsFromNewAnswersJob.create_new() # pylint: disable=line-too-long
+        stats_jobs_one_off.CleanupLargeBucketLabelsFromNewAnswersJob.enqueue(job_id) # pylint: disable=line-too-long
+        self.process_and_flush_pending_tasks()
+        return jobs.get_job_output(job_id)
+
+    def _run_migration_job(self):
+        """Start the AnswerMigrationJob to migrate answers submitted to
+        StateRuleAnswerLogModel.
+        """
+        job_id = stats_jobs_one_off.AnswerMigrationJob.create_new()
+        stats_jobs_one_off.AnswerMigrationJob.enqueue(job_id)
+        self.process_and_flush_pending_tasks()
+        return jobs.get_job_output(job_id)
+
+    def _run_migration_job_internal(self, fail_predicate=None, retry_count=0):
```

Comment:
Yes, if you look closely it's used to simulate reduce shard failures. We cannot do this without manually calling into the mapreduce job. The memcache issue is completely separate to this, though there were odd behavior differences when running the job through this versus the MR framework with regards to memcache.